internal/helm/getter/client_opts.go

```diff
@@ -69,110 +69,151 @@ func (o ClientOpts) MustLoginToRegistry() bool {
-			}
+		// Use TLS from SecretRef only if CertSecretRef is not specified (CertSecretRef takes priority)
+		if opts.TlsConfig == nil && methods.HasTLS() {
+			opts.TlsConfig = methods.TLS
+			deprecatedTLS = true
 		}
-	} else if p := obj.Spec.Provider; p != "" && p != sourcev1.GenericOCIProvider && obj.Spec.Type == sourcev1.HelmRepositoryTypeOCI && ociRepo {
-		authenticator, authErr := soci.OIDCAuth(ctx, obj.Spec.URL, obj.Spec.Provider)
-		if authErr != nil {
-			return nil, "", fmt.Errorf("failed to get credential from '%s': %w", obj.Spec.Provider, authErr)
-		}
-		hrOpts.Authenticator = authenticator
 	}
 
-	if ociRepo {
-		// Persist the certs files to the path if needed.
-		if tlsBytes != nil {
-			dir, err = os.MkdirTemp("", "helm-repo-oci-certs")
-			if err != nil {
-				return nil, "", fmt.Errorf("cannot create temporary directory: %w", err)
-			}
-			certFile, keyFile, caFile, err = storeTLSCertificateFiles(tlsBytes, dir)
-			if err != nil {
-				return nil, "", fmt.Errorf("cannot write certs files to path: %w", err)
-			}
-		}
-		loginOpt, err := registry.NewLoginOption(hrOpts.Authenticator, hrOpts.Keychain, url)
+	return deprecatedTLS, certSecret, authSecret, nil
+}
+
+// configureOCIRegistryWithSecrets sets up OCI-specific configurations using pre-fetched secrets
+func configureOCIRegistryWithSecrets(ctx context.Context, obj *sourcev1.HelmRepository, opts *ClientOpts, url string, certSecret, authSecret *corev1.Secret) (string, error) {
+	// Configure OCI authentication from authSecret if available
+	if authSecret != nil {
+		keychain, err := registry.LoginOptionFromSecret(url, *authSecret)
 		if err != nil {
-			return nil, "", err
+			return "", fmt.Errorf("failed to configure login options: %w", err)
 		}
-		if loginOpt != nil {
-			hrOpts.RegLoginOpts = []helmreg.LoginOption{loginOpt, helmreg.LoginOptInsecure(obj.Spec.Insecure)}
-			tlsLoginOpt := registry.TLSLoginOption(certFile, keyFile, caFile)
-			if tlsLoginOpt != nil {
-				hrOpts.RegLoginOpts = append(hrOpts.RegLoginOpts, tlsLoginOpt)
-			}
+		opts.Keychain = keychain
+	}
+
+	// Handle OCI provider authentication if no SecretRef
+	if obj.Spec.SecretRef == nil && obj.Spec.Provider != "" && obj.Spec.Provider != sourcev1.GenericOCIProvider {
+		authenticator, err := soci.OIDCAuth(ctx, url, obj.Spec.Provider)
+		if err != nil {
+			return "", fmt.Errorf("failed to get credential from '%s': %w", obj.Spec.Provider, err)
 		}
+		opts.Authenticator = authenticator
+	}
+
+	// Setup registry login options
+	loginOpt, err := registry.NewLoginOption(opts.Authenticator, opts.Keychain, url)
+	if err != nil {
+		return "", err
 	}
-	if deprecatedTLSConfig {
-		err = ErrDeprecatedTLSConfig
+
+	if loginOpt != nil {
+		opts.RegLoginOpts = []helmreg.LoginOption{loginOpt, helmreg.LoginOptInsecure(obj.Spec.Insecure)}
 	}
 
-	hrOpts.Insecure = obj.Spec.Insecure
+	// Handle TLS certificate files for OCI
+	var tempCertDir string
+	if opts.TlsConfig != nil {
```

Comment:
I filed https://github.com/helm/helm/issues/31075 and https://github.com/helm/helm/pull/31076 so we can stop doing this crazy dance and start leveraging our shiny new `secrets` library ;)